packages/client/src/base/model/model-change-service.ts

```diff
@@ -0,0 +1,144 @@
+/********************************************************************************
+ * Copyright (c) 2025 EclipseSource and others.
+ *
+ * This program and the accompanying materials are made available under the
+ * terms of the Eclipse Public License v. 2.0 which is available at
+ * http://www.eclipse.org/legal/epl-2.0.
+ *
+ * This Source Code may also be made available under the following Secondary
+ * Licenses when the conditions for such availability set forth in the Eclipse
+ * Public License v. 2.0 are satisfied: GNU General Public License, version 2
+ * with the GNU Classpath Exception which is available at
+ * https://www.gnu.org/software/classpath/license.html.
+ *
+ * SPDX-License-Identifier: EPL-2.0 OR GPL-2.0 WITH Classpath-exception-2.0
+ ********************************************************************************/
+import {
+    BoundsAwareViewportCommand,
+    Disposable,
+    DisposableCollection,
+    Emitter,
+    Event,
+    GModelRoot,
+    ICommand,
+    ICommandStack,
+    LazyInjector,
+    SetModelCommand,
+    SetViewportCommand,
+    TYPES,
+    UpdateModelCommand,
+    Viewport,
+    almostEquals,
+    isViewport
+} from '@eclipse-glsp/sprotty';
+import { inject, postConstruct, preDestroy } from 'inversify';
+
+/**
+ * Service that tracks changes to the model root and the viewport.
+ * Allows to register listeners that are notified when the model root or the viewport changes.
+ * The current model root can be queried at any time.
+ */
+export interface IModelChangeService {
+    /** The current model root */
+    readonly currentRoot: Readonly<GModelRoot>;
+    /**
+     * Event that is fired when the model root of the diagram changes i.e. after the `CommandStack` has processed a model update.
+     */
+    onModelRootChanged: Event<Readonly<GModelRoot>>;
+
+    /**
+     * Event that is fired when the viewport of the diagram changes i.e. after the `CommandStack` has processed a viewport update.
+     * By default, this event is only fired if the viewport was changed via a `SetViewportCommand` or `BoundsAwareViewportCommand`
+     */
+    onViewportChanged: Event<Readonly<Viewport>>;
```

Comment:
I'm wondering whether it wouldn't be more future proof to have an actual interface here instead of just `Readonly<Viewport>`? Like `ViewportChange` because we could also include the old viewport. We already have a bunch of similar events, like `SelectionChange`, `FocusChange`, `ValueChange`, `DirtyStateChange`, etc.